staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go

```diff
@@ -151,11 +210,103 @@ func WithTypeAndObjectMeta(s *schema.Structural) *schema.Structural {
 	props["metadata"] = schema.Structural{
 		Generic: schema.Generic{Type: "object"},
 		Properties: map[string]schema.Structural{
-			"name": stringType,
+			"name":         stringType,
 			"generateName": stringType,
 		},
 	}
 	result.Properties = props
 
 	return result
-}
\ No newline at end of file
+}
+
+// estimateMinSizeJSON estimates the minimum size in bytes of the given schema when serialized in JSON.
+// minLength/minProperties/minItems are not currently taken into account, so if these limits are set the
+// minimum size might be higher than what estimateMinSizeJSON returns.
+func estimateMinSizeJSON(s *schema.Structural) int64 {
+	switch s.Type {
+	case "boolean":
+		// true
+		return 4
+	case "number", "integer":
+		// 0
+		return 1
+	case "string":
+		if s.ValueValidation != nil {
+			switch s.ValueValidation.Format {
+			case "duration":
+				return minDurationSizeJSON
+			case "date":
+				return dateSizeJSON
+			case "date-time":
+				return minDatetimeSizeJSON
+			}
+		}
+		// ""
+		return 2
+	case "array":
+		// []
+		return 2
+	case "object":
+		// {}
+		objSize := int64(2)
+		// exclude required fields since the request can omit them
+		if s.ValueValidation != nil {
+			for _, propName := range s.ValueValidation.Required {
+				if prop, ok := s.Properties[propName]; ok {
+					// add 4, 2 for quotations around the property name, 1 for the colon, and 1 for a comma
+					objSize += int64(len(propName)) + estimateMinSizeJSON(&prop) + 4
+				}
+			}
+		}
+		return objSize
+	}
+	if s.XIntOrString {
+		// 0
+		return 1
+	}
+	// this code should be unreachable, but return noMaxLength just in case
+	return noMaxLength
+}
+
+// estimateMaxSizeJSON estimates the maximum number of elements that can fit in s considering request size
+// constraints.
+func estimateMaxSizeJSON(s *schema.Structural) int64 {
+	switch s.Type {
+	case "string":
+		if s.ValueValidation != nil {
+			switch s.ValueValidation.Format {
+			case "duration":
+				return maxDurationSizeJSON
+			case "date":
+				return dateSizeJSON
+			case "date-time":
+				return maxDatetimeSizeJSON
+			}
+		}
+		// subtract 2 to account for ""
+		return (maxRequestSizeBytes - 2)
+	case "array":
+		// subtract 2 to account for [ and ]
+		return (maxRequestSizeBytes - 2) / (estimateMinSizeJSON(s.Items) + 1)
+	case "object":
+		if s.AdditionalProperties != nil && s.AdditionalProperties.Structural != nil {
+			// 2 bytes for key + "" + colon + comma + smallest possible value, realistically the actual keys
+			// will all vary in length
+			// TODO(DangerOnTheRanger): is there a way to calculate how many bytes unique keys will need?
+			keyValuePairSize := estimateMinSizeJSON(s.AdditionalProperties.Structural) + 6
+			// subtract 2 to account for { and }
+			return (maxRequestSizeBytes - 2) / keyValuePairSize
+		} else {
+			// this codepath executes in the case of non-map objects,
+			// but regular objects have no concept of maxProperties
+			return noMaxLength
```

Comment:
Does this branch actually execute in the case of non-map objects? I'm having a hard time seeing how from `SchemaDeclType`, this branch isn't exercised by unit tests, and I find it confusing.

1. which of the following types of schemas is this branch for:
    * `s.AdditionalProperties != nil && s.AdditionalProperties.Structural == nil` (weird map object without defined schema for values... I've forgotten if that's possible in a structural schema)
    * `s.AdditionalProperties == nil` (normal object with properties)

2. how/where is this returned cost used? it seems like if we use `noMaxLength` in calculations anywhere, it will overrun the budget and block use of cel, but I'm not seeing how someone would adjust their schema to get it in bounds

3. if this is actually used for non-map objects, why isn't the calculation figuring out how many properties could be defined in maxRequestSizeBytes, given the declared properties? I thought we were bounding CEL expressions to paths with structural schemas with pruning enabled, so should this be something like `len(s.Properties)`?

Reply:
I think the comment is slightly misleading, unfortunately. Since non-map objects don't have `maxProperties` or some equivalent (AFAIK), they should never have this function called on them (so this codepath should be unreachable), since this function is in place for things we can iterate over in some capacity (strings, maps, arrays, etc.). I'll change the comment  to reflect that this should be unreachable in practice.

Reply:
Marking this as resolved for the time being (since I think it was just the comment at fault here), but feel free to open this back up!

Reply:
Much easier to reason about - added and fixed!